Phase 0 Networking Specifications (#763)
ethereum/eth2.0-specs: specs/networking/rpc-interface.md

ETH 2.0 Networking Spec - RPC Interface
===

# Abstract

The Ethereum 2.0 networking stack uses two modes of communication: a broadcast protocol that gossips information to interested parties via GossipSub, and an RPC protocol that retrieves information from specific clients. This specification defines the RPC protocol.

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL", NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.

# Dependencies

This specification assumes familiarity with the [Messaging](./messaging.md), [Node Identification](./node-identification), and [Beacon Chain](../core/0_beacon-chain.md) specifications.

# Specification

## Message Schemas

Message body schemas are notated like this:

```
(
    field_name_1: type
    field_name_2: type
)
```

> SSZ serialization is field-order dependent. Therefore, fields MUST be  encoded and decoded according to the order described in this document. The encoded values of each field are concatenated to form the final encoded message body. Embedded structs are serialized as Containers unless otherwise noted.

Author: Do you think it's clear enough that the field order matters if we don't mention SSZ? Technically speaking the message version (in `messaging.md`) implies the encoding format, so you're right that this is a bit redundant.

Correct, "embedded structs" means containers inside other containers.

> SSZ serialization is field-order dependent. Therefore, fields MUST be  encoded and decoded according to the order described in this document. The encoded values of each field are concatenated to form the final encoded message body. Embedded structs are serialized as Containers unless otherwise noted.

Reviewer: > Do you think it's clear enough that the field order matters if we don't mention SSZ?

If we'd replace SSZ with something that doesn't care about field order, the rest of the protocol wouldn't be affected at all. So yes, I think we don't need to mention in explicitly.

> SSZ serialization is field-order dependent. Therefore, fields MUST be  encoded and decoded according to the order described in this document. The encoded values of each field are concatenated to form the final encoded message body. Embedded structs are serialized as Containers unless otherwise noted.

Reviewer 2: With SSZ we need the schema to decode. If I get a string of bytes, I can assume an `id` in the initial bytes of the message. How do I determine if the packet is a request/response in order to decode the rest of the message?

All referenced data structures can be found in the [0-beacon-chain](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#data-structures) specification.

## `libp2p` Protocol Names

> A "Protocol Name" in `libp2p` parlance refers to a human-readable identifier `libp2p` uses in order to identify sub-protocols and stream messages of different types over the same connection. A client's supported protocol paths are negotiated by the `libp2p` stack at connection time; as such they are not part of individual message bodies.

Reviewer 3: ```suggestion
A "Protocol ID" in `libp2p` parlance refers to a human-readable identifier `libp2p` uses in order to identify sub-protocols and stream messages of different types over the same connection. Peers exchange supported protocol IDs via the `Identify` protocol upon connection. When opening a new stream, peers pin a particular protocol ID to it, and the stream remains contextualised thereafter. Since messages are sent inside a stream, they do not need to bear the protocol ID.
```

## RPC-Over-`libp2p`

> To facilitate RPC-over-`libp2p`, a single protocol path is used: `/eth/serenity/rpc/1.0.0`. Remote method calls are wrapped in a "request" structure:

Reviewer 4: what's the semantic meaning of these long version numbers?

Reviewer 5: i would imagine it's there because over time it will be bugfixed (bugfix version), updated with a commitment to being backward compatible (minor version), and updated with complete disregard for any backward compatibility, for the sake of progress (major version)

Author: Yes, the idea is that they follow semver. In practice I'd estimate that the only time we'd change these version numbers is if there was a backwards-incompatible change to the serialization/compression scheme.

See @raulk's point above re: message envelopes.

Reviewer 4: A version number can either be interpreted or not. If we rely on semver, we should specify the correct behavior for clients: consider client a that supports 1.0.0 - should it also accept 1.0.1 messages as valid automatically, or discard them? This matters for forwards compatibility.

Author: Frankly, I'm in favor of simply having integer version numbers and have a blanket statement that sub-protocol version numbers are neither forwards nor backwards compatible.

Reviewer 4: that would be my preference as well, with how the encoding and protocol looks today.

if we had a serialization format that allowed forwards/backwards compatible additions (ie adding fields) we could maybe consider two-level numbering here, where the first number would be the blanket statement, while the second would signal the version with additional fields added, which would still be compatible with previous clients.

Such an encoding is generally a good thing in the wire use case, which would be a reason to look to extensions to SSZ when used outside consensus (a super-set of SSZ for example).

Reviewer 3: +1 on integers to signal a **generation** (generation 1, generation 2...). Any reason you wouldn’t have a varint style bitmap in the HELLO message to communicate finer-grained capabilities? @arnetheduck 

I would model serialisation format and compression as part of the protocol ID. Then allow Multistream to negotiate.

Reviewer: A possible compatible change could be added message types, so I think minor version numbers could be useful in some cases.

> Any reason you wouldn’t have a varint style bitmap in the HELLO message to communicate finer-grained capabilities?

Capabilities are known from the discovery protocol/ENRs already (but we need to define what types of capabilities we need). So I don't think we need it in the HELLO message.

Reviewer 4: @raulk taking a step back, my initial understanding of the libp2p setup was that you would negotiate capabilities with discovery mainly and connect to clients you know you have common ground with - then merely verify the support by signing up to the various streams here and that each stream would be a protocol on its with libp2p dealing with the mulitplexing etc - that has changed now I see, and it looks like there's another layer of protocol negotiation within the stream to discover capabilities - that feels.. redundant, to do the same work twice, and somewhat limiting, because how does a client add a completely new message they want to test or use in some client-specific scenario (for example to establish / evaluate its usefulness) - but it seems I need to reread the newer spec.

I'll be honest though and say that I don't fully understand where the varint would go at this point with the various layers, but integers tend to be harder to negotiate than strings, in a decentralized manner - a string, you just pick one and start using it - if it becomes popular, people will avoid it. Numbers.. you need a registry and the associated maintenance.

Reviewer 3: @arnetheduck my intention isn't to propose any changes here; was just curious to hear the rationale of the Eth2.0 community re: finer-grained protocols vs. coarse-grained protocol with capabilities. We also debate this occasionally in the libp2p community ;-) [libp2p supports both patterns].

Re: semver vs. protocol generations. libp2p does not impose a specific version token (if any). When registering a handler, you can attach a [predicate](https://github.com/multiformats/go-multistream/blob/master/multistream.go#L124) to evaluate the match. So handler X could match N versions, and when receiving the callback, you can inspect the protocol pinned on the stream to infer which abilities you activate, etc.

We've traditionally used semver it in libp2p, IPFS, etc., but a few of us are not convinced of its aptness. Semver is good to convey the severity of changes in APIs, but protocol evolution is a different beast.

You generally strive to keep things backwards compatible, yet enable feature upgrades/opt-ins over time that may not be rolling cumulative, e.g. if 1.14.0 and 1.15.0 introduce feature X and Y respectively, how do I convey that a given peer supports Y but not X?

That's where protocol granularity comes into play: potentially make each feature/message/RPC a different protocol, and track "generations/revisions" of those protocols. libp2p supports that design. A few thoughts:

* We're evolving Multistream to avoid round trips when you're certain the other party supports a protocol (_selection_ vs _negotiation_), for example, via an ENR.
* If two messages have timing dependencies (can't send message B until after A), and they segregated across protocols, it may make state-tracking a bit more complicated.

---

> integers tend to be harder to negotiate than strings, in a decentralized manner - a string, you just pick one and start using it - if it becomes popular, people will avoid it.

I meant replacing semver by protocol generations, e.g. `/eth/serenity/rpc/v10`. Sorry for not being clear!

```
(
    id: uint64
    method_id: uint16
    body: Request
)
```

> and their corresponding responses are wrapped in a "response" structure:

Reviewer 6: Is it always a response, or would you send header and bodies unprompted?

Author: See below comment re: making a request portion for this.

```
(
    id: uint64
    result: Response
)
```

> If an error occurs, a variant of the response structure is returned:

Author: Ah, this is a good point. SSZ doesn't support `null` values either - let me think on this one for a little bit and come up with a solution.

> If an error occurs, a variant of the response structure is returned:

Author: Added an `is_error` boolean field. Note that with SSZ at least you can read the `is_error` field prior to the contents of the `result` via offsets. This allows clients to switch the serialized type based on the `is_error` value.

> If an error occurs, a variant of the response structure is returned:

Reviewer 4: the alternative would be to use a list - empty if there's no error, and one item if there is.

> If an error occurs, a variant of the response structure is returned:

Reviewer 4: just to be clear - when encoding or decoding ssz, there generally exists no provision for skipping fields - even if `is_error` is false, `data` must contain bytes. embedding a `StatusData` in the data field seems to go against the spirit of SSZ generally, as SSZ decoders in general expect to know the exact type of each field, thus would not fit "naturally" in "normal" ssz code.

That said, this issue stems from using SSZ in a wire protocol setting for which it is not.. great.

```
(
    id: uint64
    error: (
        code: uint16
        data: bytes
    )
)
```

The details of the RPC-Over-`libp2p` protocol are similar to [JSON-RPC 2.0](https://www.jsonrpc.org/specification). Specifically:

> 1. The `id` member is REQUIRED.

Author: Was thinking about including these just in case SSZ/whatever encoding format is used in the future start supporting `null` values.

> 2. The `id` member in the response MUST be the same as the value of the `id` in the request.

Reviewer 3: Ideally we'd define the properties of this `id`. Is it monotonically increasing? Unique within a session?

Author: According to the JSON-RPC spec it need only be unique within a session. I'll update the doc to reflect this requirement and provide some `RECOMMENDED` options.

Having an `id` parameter like this is important since it allows RPC messages to be pipelined.

Author: I've updated this. See latest changes.

3. The `method_id` member is REQUIRED.
4. The `result` member is required on success, and MUST NOT exist if there was an error.
5. The `error` member is REQUIRED on errors, and MUST NOT exist if there wasn't an error.

Structuring RPC requests in this manner allows multiple calls and responses to be multiplexed over the same stream without switching.

The "method ID" fields in the below messages refer to the `method` field in the request structure above.

> The first 1,000 values in `error.code` are reserved for system use. The following error codes are predefined:

Reviewer 5: sort of like port numbers :)

1. `0`: Parse error.
2. `10`: Invalid request.
3. `20`: Method not found.
4. `30`: Server error.

## Messages

### Hello

**Method ID:** `0`

**Body**:

```
(
    network_id: uint8
    latest_finalized_root: bytes32
    latest_finalized_epoch: uint64
    best_root: bytes32
    best_slot: uint64
)
```

Clients exchange `hello` messages upon connection, forming a two-phase handshake. The first message the initiating client sends MUST be the `hello` message. In response, the receiving client MUST respond with its own `hello` message.

Clients SHOULD immediately disconnect from one another following the handshake above under the following conditions:

> 1. If `network_id` belongs to a different chain, since the client definitionally cannot sync with this client.

Reviewer 6: You should consider spelling out network ID and chain ID as separate fields. Chain ID should be set to a fixed number "1" for ETH, and if others want to run their own chain they can change that ID.

Reviewer 7: NetworkId vs ChainId +1.
Also, message body compression algorithm indicator.
Also, upgrade paths for SSZ (I get the feeling this might change on the wire)..maybe a sorted list of serialization method preferences, the highest mutual being selected?

> 2. If the `latest_finalized_root` shared by the peer is not in the client's chain at the expected epoch. For example, if Peer 1 in the diagram below has `(root, epoch)` of `(A, 5)` and Peer 2 has `(B, 3)`, Peer 1 would disconnect because it knows that `B` is not the root in their chain at epoch 3:

Author: Cool got it, will do.

```
              Root A

              +---+
              |xxx|  +----+ Epoch 5
              +-+-+
                ^
                |
              +-+-+
              |   |  +----+ Epoch 4
              +-+-+
Root B          ^
                |
+---+         +-+-+
|xxx+<---+--->+   |  +----+ Epoch 3
+---+    |    +---+
         |
       +-+-+
       |   |  +-----------+ Epoch 2
       +-+-+
         ^
         |
       +-+-+
       |   |  +-----------+ Epoch 1
       +---+
```

> Once the handshake completes, the client with the higher `latest_finalized_epoch` or `best_slot` (if the clients have equal `latest_finalized_epoch`s) SHOULD send beacon block roots to its counterparty via `beacon_block_roots` (i.e., RPC method `10`).

Reviewer 6: This spec handles the initial sync between nodes. How are new blocks handled? Will they be pushed proactively across? Do you just push the block root?

Reviewer 6: What should they do?

Reviewer: Send beacon block roots without an explicit request. I think RPC communication should always be request response pairs, never just a response.

Author: I'm in agreement with @jannikluhn here - this should be the result of an RPC call. I'll update to fix.

Author: There will be a separate spec for gossip.

### Goodbye

**Method ID:** `1`

**Body:**

```
(
    reason: uint64
)
```

> Client MAY send `goodbye` messages upon disconnection. The reason field MUST be one of the following values:

Reviewer 8: shouldn't we still be connected after `sync finished` ? We would still need to propagate any newly proposed blocks to our peers

> Client MAY send `goodbye` messages upon disconnection. The reason field MUST be one of the following values:

Reviewer 4: generally, the standard way to sync in these kinds of "live" protocols is to start listening to broadcasts, then initiate sync.. else you'll miss packets during sync and will have to recover again.

- `1`: Client shut down.
- `2`: Irrelevant network.
- `3`: Irrelevant shard.

### Provide Beacon Block Roots

**Method ID:** `10`

**Body:**

```
# BlockRootSlot
(

>     block_root: HashTreeRoot

Reviewer 6: Is a `HashTreeRoot` a `Bytes32`?

Reviewer 9: Yes, should be updated to that or defined in this doc

    slot: uint64
)

(
    roots: []BlockRootSlot
)
```

Send a list of block roots and slots to the peer.

### Beacon Block Headers

**Method ID:** `11`

**Request Body**

```
(

>     start_root: HashTreeRoot

Reviewer: Do we need the root? It seems redundant to me, except for the case of chain reorgs which shouldn't happen frequently at sync (and even then, it's probably better to get blocks from the current chain that we'll be able to use later, instead of getting outdated ones).

Reviewer 4: ẁe need a mechanism for recovering blocks, in case something is lost or the client goes offline for a short bit and loses a few (computer went to sleep / ISP went down for 10 minutes). 

I argue in the original issue (https://github.com/ethereum/eth2.0-specs/issues/692#issuecomment-469356620) that  it's often natural to request blocks backwards for this reason: the data structure we're syncing is a singly linked list pointing backwards in time and we receive attestations and blocks that let us discover heads "naturally" by listening to the broadcasts. With a `block_root+previous_n_blocks` kind of request we can both sync and recover, and for example use attestations to discover "viable" heads to work on, from a sync or recovery perspective. Indeed, negotiating finalized epochs in the handshake is somewhat redundant in that case, albeit a nice optimization (except for the chain id) - we could equally well request blocks from the peer that gossiped us the block or attestation whose parent we're missing - they should not be gossiping attestations they have not linked to a finalized epoch of value.

Reviewer: Interesting! To summarize my understanding of your comment: Syncing forwards is safer as we can verify each block immediately when we receive it, but syncing backwards is more efficient/doesn't require additional database indexing (and I guess syncing forwards may require a negotiating phase to discover the best shared block). You're proposing to interpret the fact that I see lots of attestations on top of my sync peer's head flying around the network as evidence that their head is valid? And therefore, I'd be pretty safe syncing backwards?

That sounds reasonable. My original concern was that this requires me to know (at least some good fraction of) the validator set as otherwise my sync peer could create lots of fraudulent attestations for free that I have no chance of verifying. But I would notice this if I have at least one single honest peer (if I try to sync from them or compare the attestations coming from them).

Do you think having only a backwards sync is fine or do we need both (e.g. for highly adversarial environments, or resource constrained devices that don't participate in gossiping?).

Reviewer 4: > more efficient

In terms of network / bandwidth, I'd say it's about the same but there are some nuances:

* in forward sync, I can ask for "more" slots than already exist, potentially saving round trips - a client could use this to request "all" data at the time of request arrival. consider the following race: A sends request, a new block is produced, B receives request (similar: B starts sending response which takes time, new block is produced).
* in backward sync, one could have an `(latest-head, known_slot_number)` request ("give me the block you consider to be the head, and history back to slot N") to alleviate this race, but then the server selects the head.
* both above races are generally solved by collecting data from broadcasts while syncing (classic subscribe-then-sync pattern) - they are mainly concerns if you don't want to subscribe or want to delay subscribing.
* in forward sync, I might end up on a different branch / head than I thought I would - the request itself does not point one out

In terms of client implementations, I think of backward sync as biased to make it cheaper for the server: the server already has the data necessary - also because the head is kept hot - while the client has to keep a chain of "unknown" blocks around / can't validate eagerly. An additional rule that the response must be forward-ordered could help the client apply / validate the blocks eagerly.

The backwards sync can be seen as more passive/reactive/lazy while forward sync is more active..

> attestations on top of my sync peer's head flying around the network as evidence that their head is valid

right. the assumption rests on two premises (thanks @djrtwo!):

* honest clients will not propagate invalid data (not signed by a validator they know)
* there's a slashing condition on creating unviable attestations - there's currently no penalty to create & sign an unviable block so one can perhaps imagine a malicious group of validators creating lots of these and for example spam everyone during important periods "for free". It sounds a bit far fetched though, tbh, to be creating blocks this way - would love to hear thoughts.
* I've weak-subjectively selected an initial state that contains some validators. I'd primarily look for anything signed by those validators as another heuristic for where to start syncing (even if the validator set might have changed from there).

> Do you think having only a backwards sync is fine or do we need both (e.g. for highly adversarial environments, or resource constrained devices that don't participate in gossiping?).

I'm not sure :) I'm curious to hear feedback on this point, but here are some thoughts:
* it's important that we have a request like Hello to ask for what clients consider to be the head for non-gossiping use cases - but I think that's orthogonal to the sync direction.
* clients should be free to send that request at any time, not just during the initial negotiation phase
* direction can be different for request and response - if different, requires a slightly "smarter" server
* there's a cost for each direction, in terms of implementation. I'd start with one and look for strong motivations before implementing the other, as the returns are not that great. Either direction is sufficient, really.

Reviewer 10: > Do you think having only a backwards sync is fine or do we need both (e.g. for highly adversarial environments, or resource constrained devices that don't participate in gossiping?).

It seems reasonable to sync backwards from the latest received gossiped block (at least as an initial implementation)

> in backward sync, one could have an (latest-head, known_slot_number) request ("give me the block you consider to be the head, and history back to slot N") to alleviate this race, but then the server selects the head.

i would argue that we should drop start_slot.  if we give clients the option to request a block by either start_slot or start_root then that forces us to maintain a lookup or search mechanism for both.  if we are saying that both fields (start_slot and start_root) required to sync, then I would disagree.  we should be able to simply perform a lookup by block_root and walk the chain backwards until we reach max_headers.

Reviewer 4: >  latest received gossiped block

or even better, latest gossiped attestation

> Do we really need start_slot?

I would say that if we go with backwards sync, we should not implement forwards sync here or elsewhere unless there's a strong case for that direction. Having to implement both directions negates some of the benefits of backward sync and adds implementation surface.

It is quite possible to add forward sync in a later version of the protocol as well should it prove necessary.

Reviewer 10: > or even better, latest gossiped attestation

I can dig that

@arnetheduck or anyone else. Why do we need start_slot

    start_slot: uint64
    max_headers: uint64
    skip_slots: uint64
)
```

**Response Body:**

```
(

>     headers: []BlockHeader

Reviewer 6: I note BlockHeader is not defined in the beacon chain spec. I opened a PR to define it as a struct.

Reviewer: If it's not needed specifically for the spec, we could also just define it here.

)
```

> Requests beacon block headers from the peer starting from `(start_root, start_slot)`. The response MUST contain no more than `max_headers` headers. `skip_slots` defines the maximum number of slots to skip between blocks. For example, requesting blocks starting at slots `2` a `skip_slots` value of `2` would return the blocks at `[2, 4, 6, 8, 10]`. In cases where a slot is undefined for a given slot number, the closest previous block MUST be returned. For example, if slot `4` were undefined in the previous example, the returned array would contain `[2, 3, 6, 8, 10]`. If slot three were further undefined, the array would contain `[2, 6, 8, 10]` - i.e., duplicate blocks MUST be collapsed.

Reviewer: `skip_slots` sounds off-by-one to me: skipping 0 blocks should result in a continuous sequence.

Author: I think this makes more sense from an API perspective. Will fix.

> Requests beacon block headers from the peer starting from `(start_root, start_slot)`. The response MUST contain no more than `max_headers` headers. `skip_slots` defines the maximum number of slots to skip between blocks. For example, requesting blocks starting at slots `2` a `skip_slots` value of `2` would return the blocks at `[2, 4, 6, 8, 10]`. In cases where a slot is undefined for a given slot number, the closest previous block MUST be returned. For example, if slot `4` were undefined in the previous example, the returned array would contain `[2, 3, 6, 8, 10]`. If slot three were further undefined, the array would contain `[2, 6, 8, 10]` - i.e., duplicate blocks MUST be collapsed.

Reviewer: ```suggestion
Requests beacon block headers from the peer starting from `(start_root, start_slot)`. The response MUST contain no more than `max_headers` headers. `skip_slots` defines the maximum number of slots to skip between blocks. For example, requesting blocks starting at slots `2` a `skip_slots` value of `2` would return the blocks at `[2, 4, 6, 8, 10]`. In cases where a slot is empty for a given slot number, the closest previous block MUST be returned. For example, if slot `4` were empty in the previous example, the returned array would contain `[2, 3, 6, 8, 10]`. If slot three were further empty, the array would contain `[2, 6, 8, 10]` - i.e., duplicate blocks MUST be collapsed.
```

as technically the slots are defined, there's just no block

> The function of the `skip_slots` parameter helps facilitate light client sync - for example, in [#459](https://github.com/ethereum/eth2.0-specs/issues/459) - and allows clients to balance the peers from whom they request headers. Client could, for instance, request every 10th block from a set of peers where each per has a different starting block in order to populate block data.

Reviewer: ```suggestion
The function of the `skip_slots` parameter helps facilitate light client sync - for example, in [#459](https://github.com/ethereum/eth2.0-specs/issues/459) - and allows clients to balance the peers from whom they request headers. Clients could, for instance, request every 10th block from a set of peers where each per has a different starting block in order to populate block data.
```

### Beacon Block Bodies

**Method ID:** `12`

**Request Body:**

```
(
    block_roots: []HashTreeRoot
)
```

**Response Body:**

```
(

>     block_bodies: []BeaconBlockBody

Reviewer 6: Sorry - is there a link to what a BeaconBlockBody is?

Reviewer 9: https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#beaconblockbody

# Dependencies

This specification assumes familiarity with the [Messaging](./messaging.md), [Node Identification](./node-identification), and [Beacon Chain](../core/0_beacon-chain.md) specifications.

# Specification

## Message Schemas

Message body schemas are notated like this:

```
(
    field_name_1: type
    field_name_2: type
)
```

Embedded types are serialized as SSZ Containers unless otherwise noted.

All referenced data structures can be found in the [0-beacon-chain](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#data-structures) specification.

## `libp2p` Protocol Names

A "Protocol ID" in `libp2p` parlance refers to a human-readable identifier `libp2p` uses in order to identify sub-protocols and stream messages of different types over the same connection. Peers exchange supported protocol IDs via the `Identify` protocol upon connection. When opening a new stream, peers pin a particular protocol ID to it, and the stream remains contextualised thereafter. Since messages are sent inside a stream, they do not need to bear the protocol ID.

## RPC-Over-`libp2p`

> To facilitate RPC-over-`libp2p`, a single protocol path is used: `/eth/serenity/beacon/rpc/1.0.0`. Remote method calls are wrapped in a "request" structure:

Reviewer 11: Just curious about the reason why we have only one protocol id, instead of "one protocol for every type of request". If we have only one protocol id and use SSZ as the codec, from my understanding we might need a `method_id` prefix(either as a varint, or a fixed byte like in RLPx) for each message. This way we can read all of the bytes for a message, and then decode them according to their `method_id`

Author: `libp2p` incurs stream switching overhead every time the protocol changes. We do have to extract the method ID here, however this can be done with byte offsets in SSZ.

Reviewer 3: Let me correct an imprecision there. There’s no such thing as protocol changes inside a stream. When you open a stream, you pin a protocol to it and it stays invariant. So if you have N protocols and want to exchange messages for all of them, you open at least N streams.

Author: Am I misunderstanding the `multistream-select` spec? From my interpretation of https://github.com/multiformats/multistream-select, for one TCP connection to send/receive messages for a given protocol a new stream needs to be selected by sending a 'switch protocols' message. Is this accurate?

Author: While my nomenclature may be incorrect (apologies if it is), I'm referring to any overhead incurred on-the-wire to make two different clients speak the same protocol.

Reviewer 12: It's true that at the moment, libp2p will write the textual identifier of the protocol at the start of each opened stream and this can be considered "stream switching overhead" when making requests. The problem will be addressed in MultiStream 2.0 though, so I think we can design the wire protocol with this upgrade in mind:

https://github.com/libp2p/specs/pull/95

With Multistream 2.0, there will be an initial handshake step determining the shared protocols between two peers. Then all shared protocols will be assigned numeric IDs which will be used when opening new streams.

Reviewer 12: In other words, the ETH clients will be much simpler if `RPC-Over-libp2p` is defined to just create a new stream for each requests. We don't need to duplicate the multiplexing capabilities of libp2p in our own code.

Author: I can't tell if there is a timeline for Multistream 2.0 to be released. We can always update this spec to use Multistream 2.0 when the spec is ready, but for now I'm in favor of using the stable `libp2p` to avoid introducing implied dependencies.

3. The `id` member MUST be unique within the context of a single connection. Monotonically increasing `id`s are RECOMMENDED.
4. The `method_id` member is REQUIRED.
5. The `result` member is required on success, and MUST NOT exist if there was an error.
6. The `error` member is REQUIRED on errors, and MUST NOT exist if there wasn't an error.
7. `is_error` MUST be `true` on errors, or `false` otherwise.

Structuring RPC requests in this manner allows multiple calls and responses to be multiplexed over the same stream without switching. Note that this implies that responses MAY arrive in a different order than requests.

The "method ID" fields in the below messages refer to the `method` field in the request structure above.

The first 1,000 values in `error.code` are reserved for system use. The following error codes are predefined:

1. `0`: Parse error.
2. `10`: Invalid request.
3. `20`: Method not found.
4. `30`: Server error.

## Messages

### Hello

**Method ID:** `0`

**Body**:

```
(
    network_id: uint8
    latest_finalized_root: bytes32
    latest_finalized_epoch: uint64
    best_root: bytes32

>     best_slot: uint64

Reviewer 4: slots are based on wall time - what's the best_slot field for?

Reviewer: pretty sure this is supposed to refer to the slot of the head block. Maybe rename `best_root` and `best_slot` to `head_root` and `head_slot` (or to be even more clear `head_block_root/slot`)?

Reviewer 8: I think `head_block_root` and `head_slot` would be clearer

Clients SHOULD immediately disconnect from one another following the handshake above under the following conditions:

1. If `network_id` belongs to a different chain, since the client definitionally cannot sync with this client.
2. If the `latest_finalized_root` shared by the peer is not in the client's chain at the expected epoch. For example, if Peer 1 in the diagram below has `(root, epoch)` of `(A, 5)` and Peer 2 has `(B, 3)`, Peer 1 would disconnect because it knows that `B` is not the root in their chain at epoch 3:

```
              Root A

              +---+
              |xxx|  +----+ Epoch 5
              +-+-+
                ^
                |
              +-+-+
              |   |  +----+ Epoch 4
              +-+-+
Root B          ^
                |
+---+         +-+-+
|xxx+<---+--->+   |  +----+ Epoch 3
+---+    |    +---+
         |
       +-+-+
       |   |  +-----------+ Epoch 2
       +-+-+
         ^
         |
       +-+-+
       |   |  +-----------+ Epoch 1
       +---+
```

> Once the handshake completes, the client with the higher `latest_finalized_epoch` or `best_slot` (if the clients have equal `latest_finalized_epoch`s) SHOULD request beacon block roots from its counterparty via `beacon_block_roots` (i.e., RPC method `10`).

Reviewer 8: How would this be handled if the clients both have equal `latest_finalized_epoch` and `best_slot` ?

Reviewer 4: then you discard the client because you know it's either at genesis or providing invalid data.. finalization happens several epochs behind `best_slot` in the happy case.

Reviewer 8: @arnetheduck ah no, I was referring to if both clients have the same `best_slot` and `finalized_epoch` together . So FinalizedEpoch<sub> A </sub>  == FinalizedEpoch<sub>B</sub> and BestSlot<sub> A </sub> == BestSlot<sub> B </sub>

Reviewer 4: ah. right. I find that the situation is somewhat analogous to receiving a block whose parent is unknown to it - you have to make a very similar decision there - the information in this hello, just like in the parentless block, is essentially useless, from a trust perspective, and you need to turn to other sources. 

@djrtwo suggested that attestations might be a good heurestic as signing one carries risk for the validator that does so. The information here can, from what I can see, be used to quickly disconnect from a client, if they're saying the network is different. The rest is advisory, and you're hoping for the best.

### Goodbye

**Method ID:** `1`

**Body:**

```
(
    reason: uint64
)
```

Client MAY send `goodbye` messages upon disconnection. The reason field MUST be one of the following values:

- `1`: Client shut down.
- `2`: Irrelevant network.
- `3`: Too many peers.
- `4`: Fault/error.

### Request Beacon Block Roots

**Method ID:** `10`

**Request Body**

```
()
```

**Response Body:**

```
# BlockRootSlot
(
    block_root: bytes32
    slot: uint64
)

(

>     roots: []BlockRootSlot

Reviewer 8: Is there any size limit to the block roots in a response ? If there are no limits, I would imagine it being pretty easy to overwhelm any peer with it.

Author: Added a limit.

)
```

> Send a list of block roots and slots to the requesting peer.

Reviewer 4: which beacon block roots? known heads? all?

The function of the `skip_slots` parameter helps facilitate light client sync - for example, in [#459](https://github.com/ethereum/eth2.0-specs/issues/459) - and allows clients to balance the peers from whom they request headers. Clients could, for instance, request every 10th block from a set of peers where each per has a different starting block in order to populate block data.

### Beacon Block Bodies

**Method ID:** `12`

**Request Body:**

```
(
    block_roots: []HashTreeRoot
)
```

**Response Body:**

```
(
    block_bodies: []BeaconBlockBody
)
```

> Requests the `block_bodies` associated with the provided `block_roots` from the peer. Responses MUST return `block_roots` in the order provided in the request. If the receiver does not have a particular `block_root`, it must return a zero-value `block_body` (i.e., a `block_body` container with all zero fields).

Reviewer 4: It seems to me that when everything is going smoothly, block bodies consist of very few attestations (they should be combined by then), and a few minor items like the transfers etc. has anything looked at the numbers to see how much value there is in having separate requests for headers and bodies? Requesting headers then bodies creates additional round-trips which are a cost on its own.

## `libp2p` Protocol Names

A "Protocol ID" in `libp2p` parlance refers to a human-readable identifier `libp2p` uses in order to identify sub-protocols and stream messages of different types over the same connection. Peers exchange supported protocol IDs via the `Identify` protocol upon connection. When opening a new stream, peers pin a particular protocol ID to it, and the stream remains contextualised thereafter. Since messages are sent inside a stream, they do not need to bear the protocol ID.

## RPC-Over-`libp2p`

To facilitate RPC-over-`libp2p`, a single protocol name is used: `/eth/serenity/beacon/rpc/1`. The version number in the protocol name is neither backwards or forwards compatible, and will be incremented whenever changes to the below structures are required.

Remote method calls are wrapped in a "request" structure:

```
(
    id: uint64
    method_id: uint16
    body: Request
)
```

and their corresponding responses are wrapped in a "response" structure:

```
(
    id: uint64
    response_code: uint16

>     result: bytes

Reviewer 6: Can we just use one structure? You can tell if it's a response by comparing request IDs.

1. The `id` member is REQUIRED.
2. The `id` member in the response MUST be the same as the value of the `id` in the request.
3. The `id` member MUST be unique within the context of a single connection. Monotonically increasing `id`s are RECOMMENDED.
4. The `method_id` member is REQUIRED.
5. The `result` member is REQUIRED on success.
6. The `result` member is OPTIONAL on errors, and MAY contain additional information about the error.
7. `response_code` MUST be `0` on success.

Structuring RPC requests in this manner allows multiple calls and responses to be multiplexed over the same stream without switching. Note that this implies that responses MAY arrive in a different order than requests.

The "method ID" fields in the below messages refer to the `method` field in the request structure above.

The first 1,000 values in `response_code` are reserved for system use. The following response codes are predefined:

1. `0`: No error.
2. `10`: Parse error.
2. `20`: Invalid request.
3. `30`: Method not found.
4. `40`: Server error.

### Alternative for Non-`libp2p` Clients

> Since some clients are waiting for `libp2p` implementations in their respective languages. As such, they MAY listen for raw TCP messages on port `9000`. To distinguish RPC messages from other messages on that port, a byte prefix of `ETH` (`0x455448`) MUST be prepended to all messages. This option will be removed once `libp2p` is ready in all supported languages.

Reviewer 6: I think we should allow using a separate port. It's entirely possible a client will allow both communication modes.

Reviewer 6: Also, this paragraph should go to the top since it relates to the envelope and port 9000.

## Messages

### Hello

**Method ID:** `0`

**Body**:

```
(
    network_id: uint8

>     chain_id: uint8

Reviewer 4: the rest of the eth2 spec favors uint64 for simplicity and uniformity - what's the rationale behind `uint8` here?

Author: I think it's unlikely that we'll need to support mode than 256 chain IDs, so we can save the bytes.

Reviewer: There will be 1024 shards at least. We haven't talked about an id scheme for those yet, but to be safe I think we should go with uint64 (also considering that HELLO messages are sent only infrequently). Same for network id.

Author: Sounds good. I'll update.

    latest_finalized_root: bytes32
    latest_finalized_epoch: uint64

>     best_root: bytes32

Reviewer 4: Taking into account the backwards sync suggested elsewhere, and that we can use attestations as a (strong) heuristic that a block is valid and useful, it seems prudent to include (some) attestations here - instead of simply supplying some data like best_root that cannot be trusted anyway, a recent attestation would help the connecting client both with head / fork selection and to know with a higher degree of certainty that the root sent "makes sense" and should be downloaded.

The details of this are TBD - but probably we're looking at something like `attestations: [Attestation]` where it's up to the client to choose a representative and recent set (or none, which is also fine, because then one can listen to broadcasts).

```
              Root A

              +---+
              |xxx|  +----+ Epoch 5
              +-+-+
                ^
                |
              +-+-+
              |   |  +----+ Epoch 4
              +-+-+
Root B          ^
                |
+---+         +-+-+
|xxx+<---+--->+   |  +----+ Epoch 3
+---+    |    +---+
         |
       +-+-+
       |   |  +-----------+ Epoch 2
       +-+-+
         ^
         |
       +-+-+
       |   |  +-----------+ Epoch 1
       +---+
```

Once the handshake completes, the client with the higher `latest_finalized_epoch` or `best_slot` (if the clients have equal `latest_finalized_epoch`s) SHOULD request beacon block roots from its counterparty via `beacon_block_roots` (i.e., RPC method `10`).

### Goodbye

**Method ID:** `1`

**Body:**

```
(
    reason: uint64
)
```

Client MAY send `goodbye` messages upon disconnection. The reason field MAY be one of the following values:

- `1`: Client shut down.
- `2`: Irrelevant network.
- `3`: Fault/error.

Clients MAY define custom goodbye reasons as long as the value is larger than `1000`.

### Get Status

**Method ID:** `2`

**Request Body:**

```
(

> 	sha: bytes32

Reviewer 6: I didn't ask when this got drafted, sorry. What is sha here?

Author: The commit hash of the node.

Reviewer 6: What is that? Best root?

	user_agent: bytes
	timestamp: uint64
)
```

**Response Body:**

```
(
	sha: bytes32
	user_agent: bytes
	timestamp: uint64
)
```

Returns metadata about the remote node.

### Request Beacon Block Roots

**Method ID:** `10`

**Request Body**

```
(
	start_slot: uint64
	count: uint64
)
```

**Response Body:**

```
# BlockRootSlot
(
    block_root: bytes32
    slot: uint64
)

(
    roots: []BlockRootSlot
)
```

> Requests a list of block roots and slots from the peer. The `count` parameter MUST be less than or equal to `32768`. The slots MUST be returned in ascending slot order.

Reviewer 13: I have some questions regarding the response:

- Can you skip a slot (e.g., `1, 3, 4`)?
- Can you return less/more than the requested roots?
- Can you start at higher slot?
- Can you return a slot outside of the `start_slot + count` bounds?

Maybe _"The slots MUST be returned in ascending slot order."_ is succinct already? If this is the case we could add something like "The only requirements for `roots` are ...".

P.S. this is my first comment, thanks for making this doc!

```
(
    headers: []BeaconBlockHeader
)
```

Requests beacon block headers from the peer starting from `(start_root, start_slot)`. The response MUST contain no more than `max_headers` headers. `skip_slots` defines the maximum number of slots to skip between blocks. For example, requesting blocks starting at slots `2` a `skip_slots` value of `1` would return the blocks at `[2, 4, 6, 8, 10]`. In cases where a slot is empty for a given slot number, the closest previous block MUST be returned. For example, if slot `4` were empty in the previous example, the returned array would contain `[2, 3, 6, 8, 10]`. If slot three were further empty, the array would contain `[2, 6, 8, 10]` - i.e., duplicate blocks MUST be collapsed. A `skip_slots` value of `0` returns all blocks.

The function of the `skip_slots` parameter helps facilitate light client sync - for example, in [#459](https://github.com/ethereum/eth2.0-specs/issues/459) - and allows clients to balance the peers from whom they request headers. Clients could, for instance, request every 10th block from a set of peers where each per has a different starting block in order to populate block data.

### Beacon Block Bodies

> **Method ID:** `12`

Reviewer 6: Just like in LES, I would use different method ids for requests and responses. So it's possible for me to send you proactively blocks and headers using RPC, and you don't need to know about it in advance.

**Request Body:**

```
(
    block_roots: []HashTreeRoot
)
```

**Response Body:**

```
(

>     block_bodies: []BeaconBlockBody

Reviewer 13: To obtain blocks from a peer I expect to first request headers, then bodies. After receiving the bodies, headers are combined with the bodies and `chain.import_block(..)` (or w/e) is called.

Given that we don't have any information linking `body` to `header`, we need to iterate headers to find `header.block_body_root == hash_tree_root(body)`.

You could reduce this iteration to "all headers from that `peer_id`" rather than _all_ headers, I guess.

Nonetheless, thoughts on returning `(block_body, block_root)` to allow `O(1)` lookup of corresponding header?

Reviewer 13: Oh ignore this, I can just key the headers by `block_body_root`!
